### PR TITLE
Better test helper documentation

### LINKED ISSUE #74: Test helpers should be publically available and documented
Test helpers in some component integration tests should be exported from modules so consuming app developers can use the tooltip test helpers in their own apps.
- assertPosition
- assertVisibility
- Other options?

### LINKED ISSUE #136: document text property
text property works instead of block form but it's not documented.

### LINKED ISSUE #144: Expand test helpers documentation
The readme contains [documentation for our new pubic test helpers](https://github.com/sir-dunxalot/ember-tooltips#test-helpers). In order to encourage people to use the test helpers in their own applications, we should improve our documentation by breaking our each test helper into it's own section with an explanation of which options can be passed, and example, and an expanded description of what the helper can do and when it should be used.

### LINKED ISSUE #148: Unable to use test helpers
Following directions under Test Helpers

run `ember new tooltips-example`
then run `ember install ember-tooltips`

Copy import from example
```
import { assertTooltipVisible, assertTooltipNotVisible, assertTooltipRendered, assertTooltipNotRendered, triggerTooltipTargetEvent } from '../helpers/ember-tooltips';
```

When I run the test using `ember test --serve` I get the following error message:

```
Error: Could not find module `tooltips/tests/integration/helpers/ember-tooltips` imported from `tooltips/tests/integration/components/tooltip-example-test`
```

I could have completely missed a step but if that is the case it would be helpful to include that in the documentation about additional steps I need to do.

I can provide an example in a repository.

### PR DESCRIPTION
**Pull request by sir-dunxalot**
This PR makes the following improvements to the test suite and test helpers for this addon:

- Updates documentation and files to use absolute paths for importing test helpers (fixes #148)
- Expands test helper documentation with more explicit information and examples (fixes #144)
- Adds more documentation for the `text` property (fixes #136 and fixes #74)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -129,6 +132,16 @@ Popovers also benefit from a `hide` API made publically acessible:
 {{/popover-on-element}}
 ```
 
+## Targets
```
Comment (sir-dunxalot) [inline README.md]: I have added this section of the documentation to describe what a target is in the context of tooltips and popovers. This should make the rest of the documentation less verbose (we can just say 'target' and link back to this section for more info).

File: README.md [hunk 2]
```diff
@@ -345,6 +359,22 @@ Sets the number of pixels the tooltip will render from the target element. A hig
 {{tooltip-on-component spacing=20}}
 ```
 
+#### Text
```
Comment (sir-dunxalot) [inline README.md]: The `text` property already existed but it wasn't documented in this section.

File: addon/styles/addon.scss [hunk 3]
```diff
@@ -132,3 +132,9 @@
     }
   }
 }
+
+.ember-tooltip-or-popover-target-short-circuit {
```
Comment (sir-dunxalot) [inline addon/styles/addon.scss]: This change is [documented here](https://github.com/sir-dunxalot/ember-tooltips/pull/160/files#diff-04c6e90faac2675aa89e2176d2eec7d8R975).

Comment (sir-dunxalot) [inline addon/styles/addon.scss]: TODO: Remove this in favor of the [absolute-parent-fix](https://github.com/sir-dunxalot/ember-tooltips/tree/absolute-parent-fix) branch.

**Comment by sir-dunxalot** [comment]
Temporarily closing to make updates given improvements I've made on the [absolute-parent-fix](https://github.com/sir-dunxalot/ember-tooltips/tree/absolute-parent-fix) branch that fixes #175.

File: README.md [hunk 4]
```diff
@@ -129,6 +131,16 @@ Popovers also benefit from a `hide` API made publically acessible:
 {{/popover-on-element}}
 ```
 
+## Targets
+
+The concept of a 'target' is used through this addon. A target is the element that the tooltip of popover is attached to. Each tooltip or popvers has its own target. Interacting with this target will render and/or show the tooltip or popover.
```
Comment (a15n) [inline README.md]: `tooltip or popover` (typo fix)

File: README.md [hunk 5]
```diff
@@ -357,7 +386,7 @@ This can be useful alongside `event='none'` when you only want to toolip to show
 
 ```hbs
 {{!--Binds the tooltip visibility to the showTooltip property--}}
-{{tooltip-on-component isShown=showTooltip}}
```
Comment (a15n) [inline README.md]: why did you delete this? Also, the Controller has a `showTooltips` property. I think this is fine to leave but the Controller property should be updated to `showTooltip`.

File: README.md [hunk 6]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
+- [assertTooltipContent()](#asserttooltipcontent)
+- [assertTooltipRendered()](#asserttooltiprendered)
+- [assertTooltipNotRendered()](#asserttooltipnotrendered)
+- [assertTooltipVisible()](#asserttooltipvisible)
+- [assertTooltipNotVisible()](#asserttooltipnotvisible)
+- [assertTooltipSide()](#asserttooltipside)
+- [assertTooltipSpacing()](#asserttooltipspacing)
+- [triggerTooltipTargetEvent()](#triggertooltiptargetevent)
+
+All assert helpers require `assert` to be passed as the first param and accept a second, optional param for additional test options. For detailed usage instructions and examples, see the documentation for each test helper below.
```
Comment (a15n) [inline README.md]: `All assert helpers require `assert` to be passed as the first param and SOME accept a second, optional param for additional test options.`

File: README.md [hunk 7]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
+
+  this.render(hbs`
+    {{tooltip-on-element enableLazyRendering=true}}
+  `);
+
+  /* Tooltip won't be rendered in the DOM yet because enableLazyRendering delays the rendering until the user interacts with the target */
+
+  triggerTooltipTargetEvent($(this), 'mouseenter');
+
+  /* Now the user has interacted with the target, so the tooltip should be visible... */
```
Comment (a15n) [inline README.md]: `/* Now the user has interacted with the target, so the tooltip should be RENDERED... */`

File: README.md [hunk 8]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
+
+  triggerTooltipTargetEvent($(this), 'mouseenter');
+
+  /* Asserts that the tooltip is rendered but not shown when the user hovers over the target, which is this test's element */
+
+  assertTooltipRendered(assert);
+
+  assertTooltipNotVisible(assert);
+
+  /* We'd proabbly go on to test that another user interaction - in this case clicking this test's element - makes the tooltip visible using assertTooltipVisible() */
```
Comment (a15n) [inline README.md]: `probably`

File: README.md [hunk 9]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
+test('Example test', function(assert) {
+
+  this.render(hbs`
+    {{tooltip-on-element
+      enableLazyRendering=true
+      event='click'
+    }}
+  `);
+
+  triggerTooltipTargetEvent($(this), 'mouseenter');
```
Comment (a15n) [inline README.md]: a `mouseenter` event will not render a tooltip with a `click` event. The `assertTooltipRendered` will throw an error (we should probably fix that)

Comment (a15n) [inline README.md]: Can you double check that all these example tests would execute correctly?

File: README.md [hunk 10]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
 
-assertTooltipVisible(assert);
+  triggerTooltipTargetEvent($(this), 'mouseenter');
 
-triggerTooltipTargetEvent($tooltipTarget, 'mouseleave');
+});
+```
 
-assertTooltipNotVisible(assert);
+An options hash can be passed for more complex interactions:
```
Comment (a15n) [inline README.md]: this code doesn't show a more complex example

File: test-support/helpers/ember-tooltips.js [hunk 11]
```diff
@@ -112,6 +112,11 @@ function getTooltipAndTargetPosition(options = {}) {
   };
 }
 
+/* TODO(Duncan):
+
+Update triggerTooltipTargetEvent() to use getTooltipTargetFromBody
+and change type to side and move side into the options hash */
```
Comment (a15n) [inline test-support/helpers/ember-tooltips.js]: why would you change `type to side`? Also, I understand why we'd wnat to move `type` into the hash but that may require a deprecation period. Let's chat about this on slack.

File: README.md [hunk 12]
```diff
@@ -411,65 +440,585 @@ Four actions are available for you to hook onto through the tooltip/popover life
...
+
+  assertTooltipSpacing(assert, {
+    spacing: 35,
+  });
+});
+```
+
+#### Test helper option: `event`
+
+The name of the event that you would like to trigger on an element.
```
Comment (a15n) [inline README.md]: This may need to be updated depending on the outcome of this discussion.

https://github.com/sir-dunxalot/ember-tooltips/pull/160/files#diff-9693234231881212e47f7813b6502a69R118

File: bower.json [hunk 13]
```diff
@@ -1,12 +1,15 @@
...
+    "ember": "~1.13.0",
     "ember-cli-shims": "0.1.1",
     "ember-cli-test-loader": "0.2.2",
     "ember-qunit-notifications": "0.1.0",
     "hammer.js": "^2.0.8",
     "matchMedia": "0.2.0",
     "es5-shim": "^4.5.8"
+  },
+  "resolutions": {
+    "ember": "~1.13.0"
```
Comment (a15n) [inline bower.json]: I think you accidentally added changes to this file

File: bower.json.ember-try [hunk 14]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (a15n) [inline bower.json.ember-try]: same

File: README.md (hunk 5 above)
Comment (sir-dunxalot) [inline README.md]: The `showTooltip` property seemed like an unnecessary complication in this example when we are just trying to convey the fact that you can pass in `isShown` as a boolean.

**Comment by sir-dunxalot** [comment]
@a15n I have pushed fixes for your comments. Apologies for the stupid mistakes - I did this work on a flight. Should be good to go now.

**Comment by sir-dunxalot** [comment]
Tests seems to be failing due to an unrelated issue. I'm investigating. Restarting the Travis build did not fix the issue.

**Comment by sir-dunxalot** [comment]
@a15n This PR should shortly have a passing test suite and is ready for either a merge or final review from you. How do you want to proceed?

**Comment by a15n** [comment]
LGTM
